contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol

```diff
@@ -0,0 +1,494 @@
+    (
+      uint80 roundId,
+      int256 answer,
+      uint256 startedAt,
+      uint256 updatedAt,
+      uint80 answeredInRound
+    ) = s_arbitrumSequencerUptimeFeed.getRoundData(2);
+
+    // Validates round data
+    assertEq(roundId, 2);
+    assertEq(answer, 0);
+    assertEq(startedAt, 0);
+    assertEq(updatedAt, 0);
+    assertEq(answeredInRound, 2);
+
+    // Resets msg.sender and tx.origin
+    vm.stopPrank();
+  }
+}
+
+contract ProtectReadsOnAggregatorV2V3InterfaceFunctions is ArbitrumSequencerUptimeFeedTest {
+  /// @notice it should disallow reads on AggregatorV2V3Interface functions when consuming contract is not whitelisted
+  function test_AggregatorV2V3InterfaceDisallowReadsIfConsumingContractIsNotWhitelisted() public {
+    // Deploys a FeedConsumer contract
+    FeedConsumer feedConsumer = new FeedConsumer(address(s_arbitrumSequencerUptimeFeed));
+
+    // Sanity - consumer is not whitelisted
+    assertEq(s_arbitrumSequencerUptimeFeed.checkEnabled(), true);
+    assertEq(s_arbitrumSequencerUptimeFeed.hasAccess(address(feedConsumer), abi.encode("")), false);
+
+    // Asserts reads are not possible from consuming contract
+    vm.expectRevert("No access");
+    feedConsumer.latestAnswer();
+    vm.expectRevert("No access");
+    feedConsumer.latestRoundData();
+  }
+
+  /// @notice it should allow reads on AggregatorV2V3Interface functions when consuming contract is whitelisted
+  function test_AggregatorV2V3InterfaceAllowReadsIfConsumingContractIsWhitelisted() public {
+    // Deploys a FeedConsumer contract
+    FeedConsumer feedConsumer = new FeedConsumer(address(s_arbitrumSequencerUptimeFeed));
+
+    // Whitelist consumer
+    vm.startPrank(s_deployerAddr, s_deployerAddr);
+    s_arbitrumSequencerUptimeFeed.addAccess(address(feedConsumer));
+    vm.stopPrank();
+
+    // Sanity - consumer is whitelisted
+    assertEq(s_arbitrumSequencerUptimeFeed.checkEnabled(), true);
+    assertEq(s_arbitrumSequencerUptimeFeed.hasAccess(address(feedConsumer), abi.encode("")), true);
+
+    // Asserts reads are possible from consuming contract
+    (uint80 roundId, int256 answer, , , ) = feedConsumer.latestRoundData();
+    assertEq(feedConsumer.latestAnswer(), 0);
+    assertEq(roundId, 1);
+    assertEq(answer, 0);
+  }
+}
+
+contract GasCosts is ArbitrumSequencerUptimeFeedTest {
+  /// @notice it should consume a known amount of gas for updates @skip-coverage
+  function test_GasCosts() public {
+    // Sets msg.sender and tx.origin to a valid address
+    vm.startPrank(s_l2MessengerAddr, s_l2MessengerAddr);
+
+    // Assert initial conditions
+    uint256 timestamp = s_arbitrumSequencerUptimeFeed.latestTimestamp();
+    assertEq(s_arbitrumSequencerUptimeFeed.latestAnswer(), 0);
+
+    // Defines helper variables for measuring gas usage
+    uint256 expectedGasUsed;
+    uint256 gasStart;
+    uint256 gasFinal;
+
+    // measures gas used for no update
+    expectedGasUsed = 5507; // TODO: used to be 28300
```

Comment:
That sounds like a good improvement :p

Reply:
Generally you don't need specific gas tests in Foundry, as all tests will be part of the snapshot anyway. So you will always have full gas coverage on all tests. You can however still do more accurate gas measurements with 

```
vm.pauseGasMetering();
vm.resumeGasMetering();
```

those calls stop and start gas measurements for when you are doing setup calls/logic in the test. You then only turn on measuring for the single call you want to measure. 

Take this test as example


```
  function testTransmitSuccess_gas() public {
    vm.pauseGasMetering();
    bytes32[3] memory reportContext = [s_configDigest, s_configDigest, s_configDigest];

    changePrank(s_valid_transmitters[0]);
    vm.resumeGasMetering();
    s_OCR2Base.transmit(reportContext, REPORT, s_rs, s_ss, s_rawVs);
  }
```


You could even load storage vars into memory before the call to be more exact